[ATen] Add CPU fp16 support for nll_loss and cross_entropy_loss

etaf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123256

Add CPU FP16 support for nll_loss and cross_entropy_loss.
Resolve issue #123328.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

etaf committed:
[ATen] Add CPU fp16 support for nll_loss_out_frame

Signed-off-by: xinan.lin <xinan.lin@intel.com>

[ghstack-poisoned]

jgong5 reviewed (approved):
Do we have UT?

etaf commented:
> Do we have UT?

I'v tried, but not find any existing test case for `nll_loss_out_frame` in Pytorch.

EikanWang reviewed (approved):
Could you help add UT to `test_ops.py`? So it could help XPU as well.

etaf committed:
Update on "[ATen] Add CPU fp16 support for nll_loss_out_frame"


Fix issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf commented:
> Do we have UT?

UT added.

EikanWang reviewed (approved):
LGTM.

EikanWang commented:
@etaf , please fix the failures in terms of the sample input.

jgong5 reviewed (changes requested):
UT failing...

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss_out_frame"


Fix issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss_out_frame"


Fix issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss_out_frame"


Fix issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss_out_frame"


Fix issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss_out_frame"


Fix issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss_out_frame"


Fix issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss_out_frame"


Fix issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss and cross_entropy_loss"


Resolve issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss and cross_entropy_loss"


Resolve issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss and cross_entropy_loss"


Resolve issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[WIP][ATen] Add CPU fp16 support for nll_loss and cross_entropy_loss"


Resolve issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

EikanWang commented:
@jgong5 , may I know if you have any additional comments?

etaf commented:
@mruberry @kulinseth @BowenBao @abock @thiagocrepaldi @wschin 
Hello, can any one help to have a review?

malfet commented:
@pytorchbot merge -i

etaf committed:
Update on "[ATen] Add CPU fp16 support for nll_loss and cross_entropy_loss"


Add CPU FP16 support for nll_loss and cross_entropy_loss.
Resolve issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf committed:
Update on "[ATen] Add CPU fp16 support for nll_loss and cross_entropy_loss"


Add CPU FP16 support for nll_loss and cross_entropy_loss.
Resolve issue #123328.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

etaf commented:
@pytorchbot  merge

malfet commented on test/test_mps.py:
```diff
@@ -987,7 +989,10 @@ def mps_ops_modifier(ops):
-        'nn.functional.binary_cross_entropy': [torch.float16]
+        'nn.functional.binary_cross_entropy': [torch.float16],
+
+        'nn.functional.nll_loss': [torch.float16],
```
@etaf in the future, please add a comment why are you adding a function to a skip-list (and create an issue)
I though the reason for the skip were that it crashed the runtime, but looks like it just it needed a precision tolerance tweak

etaf replied:
@malfet thanks for your reminder,  I'll take care of it next time。
